Autotools build system

mkoeppe opened the pull request:
Here's the autotools build system that I started while visiting Osnabrück. It is now complete (including SCIP and CoCoALib interfacing, building nmzIntegrate, etc.). 
This branch does not yet include  the 'outout' branch -- instead `make check` copies the input files from the testsuite into the build directory.

Please test. See the updated `source/INSTALL` file for instructions. It works here on Mac OS X (and it seems that the issue mentioned in https://www.normaliz.uni-osnabrueck.de/download/release-notes/ regarding the Mac OS X is fixed with this build). 

`make distcheck` builds a new distribution. It does not yet do anything regarding the PDF documentation and about jNormaliz.

mkoeppe commented:
I've made some more changes in the configure script so that it is less strict about the required C++ 11 support (normaliz just uses a few features at the moment). So now the build goes through with both clang++ and the old g++ that are used on Travis CI. So there is a good chance that it will also go through on many older Linux distributions. (In addition, I have tested it on Mac OS X.)

I have made changes to the `.travis.yml` file so that both the CMake build system and the new autotools build system are tested. In addition, `make distcheck` is tested to make sure that `make dist` succeeds in building a working normaliz distribution.

What needs to be tested manually is: SCIP, CoCoA. This is not done by the Travis script.

mkoeppe commented:
OK, now there's also a Travis CI build for `nmzIntegrate` (which first downloads and builds CoCoALib). See `.travis-build.sh`.

mkoeppe commented:
It now detects libscip in various forms, however when building with SCIP, I get:
```
$ source/normaliz  -n test/test-n/hickerson-16
Segmentation fault: 11
```
This is with scipoptsuite 3.2.1; this needs checking.

RichardSieg commented:
It works perfectly fine with scipoptsuite 3.2.0. I will check with 3.2.1 on my computer. The ZIB team told me that nothing has changed that would concern our code. What is gdb saying?

mkoeppe commented:
> What is gdb saying?

```
Program received signal SIGSEGV, Segmentation fault.
0x00007fff94ba62da in stack_not_16_byte_aligned_error () from /usr/lib/system/libdyld.dylib
(gdb) where
#0  0x00007fff94ba62da in stack_not_16_byte_aligned_error () from /usr/lib/system/libdyld.dylib
#1  0x0000000103509880 in ?? ()
#2  0x00000001010ac590 in ?? () from /Users/mkoeppe/cvs/sage/local/lib/libsoplex-2.2.1.darwin.x86_64.gnu.opt.dylib
#3  0x00000000000005eb in ?? ()
#4  0x0000000100e41b47 in soplex::CLUFactor::solveRight(double*, double*) ()
   from /Users/mkoeppe/cvs/sage/local/lib/libsoplex-2.2.1.darwin.x86_64.gnu.opt.dylib
#5  0x0000000100e9aada in soplex::SLUFactor::solveRight(soplex::VectorBase<double>&, soplex::VectorBase<double> const&) ()
   from /Users/mkoeppe/cvs/sage/local/lib/libsoplex-2.2.1.darwin.x86_64.gnu.opt.dylib
#6  0x0000000101027c2a in soplex::SPxSolver::init() () from /Users/mkoeppe/cvs/sage/local/lib/libsoplex-2.2.1.darwin.x86_64.gnu.opt.dylib
#7  0x0000000101011199 in soplex::SPxSolver::solve() () from /Users/mkoeppe/cvs/sage/local/lib/libsoplex-2.2.1.darwin.x86_64.gnu.opt.dylib
#8  0x0000000100c4ab5a in SPxSCIP::doSolve(bool) () from /Users/mkoeppe/cvs/sage/local/lib/liblpispx-3.2.1.darwin.x86_64.gnu.opt.dylib
#9  0x0000000100c4b514 in SPxSCIP::solve() () from /Users/mkoeppe/cvs/sage/local/lib/liblpispx-3.2.1.darwin.x86_64.gnu.opt.dylib
#10 0x0000000100c36386 in spxSolve(SCIP_LPi*, soplex::SPxSolver::Representation, soplex::SPxSolver::Type) ()
   from /Users/mkoeppe/cvs/sage/local/lib/liblpispx-3.2.1.darwin.x86_64.gnu.opt.dylib
#11 0x0000000100800aba in lpAlgorithm () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.opt.dylib
#12 0x000000010068a471 in lpSolve () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.opt.dylib
#13 0x000000010068c5db in SCIPlpSolveAndEval () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.opt.dylib
#14 0x0000000100783478 in propAndSolve.isra.16 () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.opt.dylib
#15 0x00000001007855c7 in SCIPsolveCIP () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.opt.dylib
#16 0x0000000100757d76 in SCIPsolve () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.opt.dylib
#17 0x000000010014f340 in std::vector<long long, std::allocator<long long> > libnormaliz::opt_sol<long long>(Scip*, libnormaliz::Matrix<long long> const&, libnormaliz::Matrix<long long> const&, std::vector<long long, std::allocator<long long> > const&) ()
#18 0x000000010014ff51 in void libnormaliz::bottom_points_inner<long long>(std::list<std::vector<long long, std::allocator<long long> >, std::allocator<std::vector<long long, std::allocator<long long> > > > const&, Scip*, libnormaliz::Matrix<long long>&, std::list<std::vector<long long, std::allocator<long long> >, std::allocator<std::vector<long long, std::allocator<long long> > > >&, std::vector<libnormaliz::Matrix<long long>, std::allocator<libnormaliz::Matrix<long long> > >&, std::vector<libnormaliz::Matrix<long long>, std::allocator<libnormaliz::Matrix<long long> > >&, long) ()
#19 0x000000010014a71e in libnormaliz::bottom_points<long long> () at libnormaliz/bottom.cpp:208
#20 0x0000000100b8a8af in GOMP_parallel () from /Users/mkoeppe/cvs/sage/local/lib/libgomp.1.dylib
#21 0x000000010014c1a4 in libnormaliz::bottom_points<long long> (new_points=..., gens=..., grading_=..., app_level=<optimized out>, 
    recursion_depth=<optimized out>) at libnormaliz/bottom.cpp:167
#22 0x0000000100135070 in libnormaliz::SimplexEvaluator<long long>::Simplex_parallel_evaluation() ()
#23 0x00000001000c2310 in libnormaliz::Full_Cone<long long>::evaluate_large_simplex (this=<optimized out>, j=<optimized out>, lss=<optimized out>)
    at libnormaliz/full_cone.cpp:2446
#24 0x00000001000c24de in libnormaliz::Full_Cone<long long>::evaluate_large_simplices (this=<optimized out>) at libnormaliz/full_cone.cpp:2409
#25 0x00000001000c2743 in libnormaliz::Full_Cone<long long>::primal_algorithm_finalize (this=<optimized out>) at libnormaliz/full_cone.cpp:2625
#26 0x00000001000c1127 in libnormaliz::Full_Cone<long long>::primal_algorithm (this=<optimized out>) at libnormaliz/full_cone.cpp:2591
#27 0x00000001000c1368 in libnormaliz::Full_Cone<long long>::compute (this=<optimized out>) at libnormaliz/full_cone.cpp:2926
#28 0x000000010013b65a in void libnormaliz::Cone<__gmp_expr<__mpz_struct [1], __mpz_struct [1]> >::compute_inner<long long>(libnormaliz::ConeProperties&) ()
#29 0x00000001000e1f5b in libnormaliz::Cone<__gmp_expr<__mpz_struct [1], __mpz_struct [1]> >::compute (this=<optimized out>, 
    ToCompute=<error reading variable: Could not find the frame base for "libnormaliz::Cone<__gmp_expr<__mpz_struct [1], __mpz_struct [1]> >::compute(libnormaliz::ConeProperties)".>) at libnormaliz/cone.cpp:1482
#30 0x0000000100018523 in int process_data<__gmp_expr<__mpz_struct [1], __mpz_struct [1]> >(OptionsHandler&) ()
#31 0x00000001001545ac in main ()

```

mkoeppe commented:
Turns out this error has nothing to do with normaliz. Something's wrong with SCIP or SoPlex on my computer (or more general its shared library build on Mac).

mkoeppe commented:
OK, I now have a working SCIP installation (getting the shared library build working on Mac OS X was tricky - see https://trac.sagemath.org/ticket/21094); but now I see a different error with normaliz when I configure it with both SCIP and OpenMP.

```
(sage-sh) mkoeppe@egret:Normaliz$ gdb --args source/normaliz  -n test/test-n/hickerson-16
GNU gdb (GDB) 7.8 [...]
This GDB was configured as "x86_64-apple-darwin15.4.0". [...]
Reading symbols from source/normaliz...done.
(gdb) run
Starting program: /Users/mkoeppe/s/sage/Normaliz/source/normaliz -n test/test-n/hickerson-16
[src/scip/paramset.c:1455] ERROR: [src/scip/paramset.c:1455] ERROR: [src/scip/paramset.c:1455] ERROR: Tried to free unknown pointer <0x102a02840>.
Tried to free unknown pointer <0x101b1e960>.
Tried to free unknown pointer <0x101b214f0>.
[src/scip/paramset.c:1455] ERROR: Tried to free unknown pointer <0x10381fe00>.
[src/scip/paramset.c:1455] ERROR: Tried to free unknown pointer <0x105000a00>.
[src/scip/paramset.c:1455] ERROR: Tried to free unknown pointer <0x106000000>.
[src/scip/paramset.c:1455] ERROR: Tried to free unknown pointer <0x105000e00>.
[src/scip/set.c:3895] ERROR: Tried to free unknown pointer <0x102907f90>.
[src/scip/set.c:3206] ERROR: Tried to free unknown pointer <0x101b28cb0>.
[src/scip/set.c:3203] ERROR: Tried to free unknown pointer <0x101c253c0>.
[src/scip/set.c:3204] ERROR: Tried to free unknown pointer <0x101c25420>.
[src/scip/set.c:3205] ERROR: Tried to free unknown pointer <0x101c25480>.
[src/scip/set.c:3206] ERROR: Tried to free unknown pointer <0x101c254e0>.
[src/scip/set.c:3203] ERROR: Tried to free unknown pointer <0x102808af0>.
[src/scip/set.c:3203] ERROR: Tried to free unknown pointer <0x102a1e3a0>.
[src/scip/dialog.c:801] ERROR: Tried to free unknown pointer <0x101b4c480>.
[src/scip/set.c:4152] ERROR: Tried to free unknown pointer <0x101f45ac0>.
[src/scip/dialog.c:801] ERROR: Tried to free unknown pointer <0x102c0db90>.
[src/scip/dialog.c:801] ERROR: Tried to free unknown pointer <0x102910270>.
[src/scip/set.c:4152] ERROR: Tried to free unknown pointer <0x102808c70>.
[src/scip/dialog.c:801] ERROR: Tried to free unknown pointer <0x102909d20>.
[src/scip/set.c:4152] ERROR: Tried to free unknown pointer <0x102810e70>.
[src/scip/set.c:4152] ERROR: Tried to free unknown pointer <0x102811810>.
[src/scip/dialog.c:801] ERROR: Tried to free unknown pointer <0x101f48080>.
[src/scip/dialog.c:801] ERROR: Tried to free unknown pointer <0x102b1e2e0>.
[src/scip/set.c:4152] ERROR: Tried to free unknown pointer <0x102c175b0>.
[src/scip/set.c:4152] ERROR: Tried to free unknown pointer <0x102918d60>.
[src/scip/dialog_default.c:4893] ERROR: Tried to free unknown pointer <0x101b50d20>.
[src/scip/dialog_default.c:4893] ERROR: Tried to free unknown pointer <0x101c516c0>.
[src/scip/dialog_default.c:4893] ERROR: Tried to free unknown pointer <0x102c17d90>.
[src/scip/dialog_default.c:4893] ERROR: Tried to free unknown pointer <0x102b1b290>.
[src/scip/dialog_default.c:4893] ERROR: [src/scip/dialog_default.c:4893] ERROR: [src/scip/dialog_default.c:4893] ERROR: Tried to free unknown pointer <0x101b50d20>.
Tried to free unknown pointer <0x102814340>.
Tried to free unknown pointer <0x101f55290>.
[src/scip/dialog.c:801] ERROR: [src/scip/dialog_default.c:4893] ERROR: [src/scip/dialog_default.c:4893] ERROR: [src/scip/dialog_default.c:4893] ERROR: Tried to free unknown pointer <0x101c59b90>.
Tried to free unknown pointer <0x1028135d0>.
Tried to free unknown pointer <0x102b1b290>.
[New Thread 0x1313 of process 49202]
[New Thread 0x1403 of process 49202]
[New Thread 0x1503 of process 49202]
[New Thread 0x1603 of process 49202]
[New Thread 0x1703 of process 49202]
[New Thread 0x1803 of process 49202]
[New Thread 0x1903 of process 49202]

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x1313 of process 49202]
0x000000010146e030 in removeMemlistEntry () from /Users/mkoeppe/cvs/sage/local/lib/liblpispx-3.2.1.darwin.x86_64.gnu.dbg.dylib
(gdb) where
#0  0x000000010146e030 in removeMemlistEntry () from /Users/mkoeppe/cvs/sage/local/lib/liblpispx-3.2.1.darwin.x86_64.gnu.dbg.dylib
#1  0x000000010146f0fd in BMSreallocMemoryArray_call () from /Users/mkoeppe/cvs/sage/local/lib/liblpispx-3.2.1.darwin.x86_64.gnu.dbg.dylib
#2  0x0000000100b78c9c in SCIPdialogAddEntry () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.dbg.dylib
#3  0x0000000100ca629e in SCIPaddDialogEntry () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.dbg.dylib
#4  0x0000000100a2296c in addSetParamDialog () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.dbg.dylib
#5  0x0000000100a22701 in addSetParamDialog () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.dbg.dylib
#6  0x0000000100a2c252 in SCIPincludeDialogDefaultSet () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.dbg.dylib
#7  0x0000000100a31fb4 in SCIPincludeDialogDefault () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.dbg.dylib
#8  0x00000001008a35ae in SCIPincludeConshdlrCountsols () from /Users/mkoeppe/cvs/sage/local/lib/libscip-3.2.1.darwin.x86_64.gnu.dbg.dylib
#9  0x0000000100d01ef4 in SCIPincludeDefaultPlugins (scip=0x101b19990) at src/scip/scipdefplugins.c:41
#10 0x00000001001ae50a in libnormaliz::bottom_points<long long> () at libnormaliz/bottom.cpp:178
#11 0x00000001013a754d in gomp_thread_start () from /Users/mkoeppe/cvs/sage/local/lib/libgomp.1.dylib
#12 0x00007fff92e9e99d in _pthread_body () from /usr/lib/system/libsystem_pthread.dylib
#13 0x00007fff92e9e91a in _pthread_start () from /usr/lib/system/libsystem_pthread.dylib
#14 0x00007fff92e9c351 in thread_start () from /usr/lib/system/libsystem_pthread.dylib
#15 0x0000000000000000 in ?? ()
```

These errors go away if I configure normaliz with `--disable-openmp`.
Do I need to compile SCIP specially to be safe for multithreaded operation?

RichardSieg commented:
Hi! Usually we do not have any problems with scip and openmp and it works perfectely fine on my computer. I will check your configuration.

mkoeppe commented:
SCIP support now also works with a non-installed scipoptsuite without any build patches.
I have updated the installation instructions, which warn against using SCIP's broken shared library build; and added a Travis CI build "autotools-scip".
Richard, Winfried - please test on your machines.

RichardSieg commented:
John was pointing out, that there might be a superfluous or void line in configure.ac:

AX_CXX_COMPILE_STDCXX(0x, , mandatory)

The first parameter may be a 11 or 14 and is set to 11 otherwise. Is that correct?

mkoeppe commented:
I provide a patched macro AX_CXX_COMPILE_STDCXX that supports the parameter
0x.
See m4/ax_cxx_compile_stdcxx.m4 for the definition of this macro.

This was necessary for the old compilers that are found on Travis. They
don't support full C++ 11.



On Thu, Sep 8, 2016 at 1:54 AM, Richard Sieg <notifications@github.com>
wrote:

> John was pointing out, that there might be a superfluous or void line in
> configure.ac:
>
> AX_CXX_COMPILE_STDCXX(0x, , mandatory)
>
> The first parameter may be a 11 or 14 and is set to 11 otherwise. Is that
> correct?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Normaliz/Normaliz/pull/102#issuecomment-245535083>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH9WhdsKulY3P4ZxNvNIZ7ajxn2cPCWnks5qn805gaJpZM4JTVk_>
> .
>



-- 
Dr. Matthias Koeppe . . . . . . . . . . .
http://www.math.ucdavis.edu/~mkoeppe
Professor, Mathematics
Chair, Applied Mathematics (Graduate Group) . .
http://appliedmath.ucdavis.edu

RichardSieg commented:
Okay, thanks for the clarification :)
